sympy/matrices/tests/test_matrices.py

```diff
@@ -1958,15 +1958,7 @@ def test_diff():
     assert A.diff(a) == MutableDenseMatrix([[0, 0], [0, 0]])
 
     B = ImmutableDenseMatrix([a, b])
-    assert A.diff(B) == Array(
-        [[[
-            [0,0],
-            [0,0]
-        ]],
-        [[
-            [0,0],
-            [0,0]
-        ]]])
+    assert A.diff(B) == A.zeros(2)
```

Comment:
Was this an incorrect result before?

Reply:
If something is detected as having a zero result it was returned like this:
https://github.com/sympy/sympy/blob/03ffa11604bcd59a75e3bb6acebe9e34416f8c65/sympy/core/function.py#L1273-L1277
The detection has been improved so it now gives this quick exit result. If it is important to keep the previous behavior then an additional `if` in the if-block will need to be added that tests whether a matrix is in the 'wrt' variables.

Reply:
I'm not sure what the result is supposed to be. Is differentiating a 3x3 matrix wrt to a 2x1 matrix supposed to give a 2x1? Wouldn't it give a 3x3x2?

Reply:
Previous result is a 2x1x2x2. Not sure where that comes from:
```
In [1]: import numpy as np

In [2]: a=np.array(
   ...:         [[[
   ...:             [0,0],
   ...:             [0,0]
   ...:         ]],
   ...:         [[
   ...:             [0,0],
   ...:             [0,0]
   ...:         ]]])
   ...:         

In [3]: a.shape
Out[3]: (2, 1, 2, 2)
```

Reply:
> Not sure where that comes from

nor I.

Reply:
CC @Upabjojr

Reply:
I think the shape should be 2x1x3x3.

Reply:
> shape should be 2x1x3x3.

Please open a new issue if this is the case. I won't be adding more to this PR.